iml-services/iml-device/src/virtual_device/util.rs

```diff
@@ -0,0 +1,153 @@
+use device_types::devices::{
+    Device, LogicalVolume, MdRaid, Mpath, Partition, Root, ScsiDevice, VolumeGroup, Zpool,
+};
+use im::OrdSet;
+use std::path::PathBuf;
+
+pub fn children_owned(d: Device) -> OrdSet<Device> {
+    match d {
+        Device::Root(dd) => dd.children,
+        Device::ScsiDevice(dd) => dd.children,
+        Device::Partition(dd) => dd.children,
+        Device::MdRaid(dd) => dd.children,
+        Device::Mpath(dd) => dd.children,
+        Device::VolumeGroup(dd) => dd.children,
+        Device::LogicalVolume(dd) => dd.children,
+        Device::Zpool(dd) => dd.children,
+        Device::Dataset(_) => OrdSet::new(),
+    }
+}
+
+pub fn children_mut(d: &mut Device) -> Option<&mut OrdSet<Device>> {
+    match d {
+        Device::Root(dd) => Some(&mut dd.children),
+        Device::ScsiDevice(dd) => Some(&mut dd.children),
+        Device::Partition(dd) => Some(&mut dd.children),
+        Device::MdRaid(dd) => Some(&mut dd.children),
+        Device::Mpath(dd) => Some(&mut dd.children),
+        Device::VolumeGroup(dd) => Some(&mut dd.children),
+        Device::LogicalVolume(dd) => Some(&mut dd.children),
+        Device::Zpool(dd) => Some(&mut dd.children),
+        Device::Dataset(_) => None,
+    }
+}
+
+pub fn children(d: &Device) -> Option<&OrdSet<Device>> {
+    match d {
+        Device::Root(dd) => Some(&dd.children),
+        Device::ScsiDevice(dd) => Some(&dd.children),
+        Device::Partition(dd) => Some(&dd.children),
+        Device::MdRaid(dd) => Some(&dd.children),
+        Device::Mpath(dd) => Some(&dd.children),
+        Device::VolumeGroup(dd) => Some(&dd.children),
+        Device::LogicalVolume(dd) => Some(&dd.children),
+        Device::Zpool(dd) => Some(&dd.children),
+        Device::Dataset(_) => None,
+    }
+}
+
+pub fn compare_selected_fields(a: &Device, b: &Device) -> bool {
+    selected_fields(a) == selected_fields(b)
+}
+
+fn selected_fields(d: &Device) -> Device {
+    match d {
```

Comment:
There is a lot of cloning happening here.

Reply:
I can rewrite it w/o cloning. Does it still need to compare all of fields (except for `major`, `minor`, `paths` and `devpath`) or just serials?

Reply:
So with provisional version that only compares serials and does no cloning for comparisons, that part of flamegraph went away completely. But the interesting thing is that total time spent in the loop didn't decrease as much (average time is about 300ms still).

I think I need to test with a no-op instead of actual device resolution because this looks like waiting for IO.

Reply:
It's dependant on device type but there should be something akin to a unique id for each variant.